File: ts/src/evidence/tracker.ts
/** Evidence access tracking (AC-504). */

import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import type { EvidenceWorkspace } from "./workspace.js";

const ACCESS_LOG_FILENAME = "evidence_access_log.json";

export function recordAccess(
  workspace: EvidenceWorkspace,
  artifactId: string,
): void {
  if (!workspace.accessedArtifacts.includes(artifactId)) {
    workspace.accessedArtifacts.push(artifactId);
  }
}

export function saveAccessLog(workspace: EvidenceWorkspace): void {
  const logPath = join(workspace.workspaceDir, ACCESS_LOG_FILENAME);
  writeFileSync(
    logPath,
    JSON.stringify({ accessed: workspace.accessedArtifacts }, null, 2),
    "utf-8",
  );
}

export function loadAccessLog(workspaceDir: string): string[] {
  const logPath = join(workspaceDir, ACCESS_LOG_FILENAME);
  if (!existsSync(logPath)) return [];
  try {
    const data = JSON.parse(readFileSync(logPath, "utf-8")) as {
      accessed?: string[];
    };
    return data.accessed ?? [];
  } catch {
    return [];
  }
}

export function computeUtilization(workspace: EvidenceWorkspace): {
  totalArtifacts: number;
  accessedCount: number;
  utilizationPercent: number;
  byKind: Record<string, { total: number; accessed: number }>;
} {
  const total = workspace.artifacts.length;
  const accessed = workspace.accessedArtifacts.length;
  const pct = total > 0 ? Math.round((accessed / total) * 1000) / 10 : 0;

  const accessedSet = new Set(workspace.accessedArtifacts);
  const byKind: Record<string, { total: number; accessed: number }> = {};
  for (const a of workspace.artifacts) {
    if (!byKind[a.kind]) byKind[a.kind] = { total: 0, accessed: 0 };
    byKind[a.kind].total++;
    if (accessedSet.has(a.artifactId)) byKind[a.kind].accessed++;
  }

  return {
    totalArtifacts: total,
    accessedCount: accessed,
    utilizationPercent: pct,
    byKind,
  };
}

File: autocontext/src/autocontext/evidence/tracker.py
"""Evidence access tracking (AC-504)."""

from __future__ import annotations

import json
from collections import Counter
from pathlib import Path
from typing import Any

from autocontext.evidence.workspace import EvidenceWorkspace

_ACCESS_LOG_FILENAME = "evidence_access_log.json"


def record_access(workspace: EvidenceWorkspace, artifact_id: str) -> None:
    """Record that an artifact was consulted. Deduplicates."""
    if artifact_id not in workspace.accessed_artifacts:
        workspace.accessed_artifacts.append(artifact_id)


def save_access_log(workspace: EvidenceWorkspace) -> None:
    """Persist the access log as JSON alongside the workspace."""
    log_path = Path(workspace.workspace_dir) / _ACCESS_LOG_FILENAME
    log_path.write_text(
        json.dumps({"accessed": workspace.accessed_artifacts}, indent=2),
        encoding="utf-8",
    )


def load_access_log(workspace_dir: str) -> list[str]:
    """Load the access log from a workspace directory."""
    log_path = Path(workspace_dir) / _ACCESS_LOG_FILENAME
    if not log_path.exists():
        return []
    try:
        data = json.loads(log_path.read_text(encoding="utf-8"))
        if not isinstance(data, dict):
            return []
        accessed = data.get("accessed", [])
        if not isinstance(accessed, list):
            return []
        return [artifact_id for artifact_id in accessed if isinstance(artifact_id, str)]
    except (json.JSONDecodeError, OSError):
        return []


def compute_utilization(workspace: EvidenceWorkspace) -> dict[str, Any]:
    """Return utilization stats."""
    total = len(workspace.artifacts)
    accessed = len(workspace.accessed_artifacts)
    pct = round(accessed / total * 100, 1) if total > 0 else 0.0

    kind_counts: Counter[str] = Counter(a.kind for a in workspace.artifacts)
    accessed_set = set(workspace.accessed_artifacts)
    kind_accessed: Counter[str] = Counter(a.kind for a in workspace.artifacts if a.artifact_id in accessed_set)
    by_kind: dict[str, dict[str, int]] = {}
    for kind in kind_counts:
        by_kind[kind] = {
            "total": kind_counts[kind],
            "accessed": kind_accessed.get(kind, 0),
        }

    return {
        "total_artifacts": total,
        "accessed_count": accessed,
        "utilization_percent": pct,
        "by_kind": by_kind,
    }
